### PR TITLE
keep the etag information of the object

### PR DESCRIPTION
**Pull request by zhijian-pro**
display the etag information of the object

### CODE REVIEW

File: cmd/gateway.go [hunk 1]
```diff
@@ -82,6 +82,10 @@ func gatewayFlags() *cli.Command {
 		&cli.BoolFlag{
 			Name:  "multi-buckets",
 			Usage: "use top level of directories as buckets",
+		},
+		&cli.BoolFlag{
+			Name:  "display-etag",
```
Comment (davies) [inline cmd/gateway.go:87]: keep-etag or save-etag

File: pkg/gateway/gateway.go [hunk 2]
```diff
@@ -467,10 +468,17 @@ func (n *jfsObjects) CopyObject(ctx context.Context, srcBucket, srcObject, dstBu
 		err = jfsToObjectErr(ctx, eno, dstBucket, dstObject)
 		return
 	}
+
+	var etag []byte
+	if n.displayEtag {
+		etag, _ = n.fs.GetXattr(mctx, src, partEtag)
```
Comment (davies) [inline pkg/gateway/gateway.go:474]: Some of file may not have etag, ignore it

File: pkg/gateway/gateway.go [hunk 3]
```diff
@@ -530,13 +538,18 @@ func (n *jfsObjects) GetObjectInfo(ctx context.Context, bucket, object string, o
 		err = jfsToObjectErr(ctx, os.ErrNotExist, bucket, object)
 		return
 	}
+	var etag []byte
+	if n.displayEtag {
+		etag, _ = n.fs.GetXattr(mctx, n.path(bucket, object), partEtag)
```
Comment (davies) [inline pkg/gateway/gateway.go:543]: Return etag only when it's saved

File: pkg/gateway/gateway.go [hunk 4]
```diff
@@ -856,6 +885,13 @@ func (n *jfsObjects) CompleteMultipartUpload(ctx context.Context, bucket, object
 
 	// Calculate s3 compatible md5sum for complete multipart.
 	s3MD5 := minio.ComputeCompleteMultipartMD5(parts)
+	if n.keepEtag {
+		eno = n.fs.SetXattr(mctx, name, s3Etag, []byte(s3MD5), 0)
+		if eno != 0 {
+			logger.Errorf("set xattr error, path: %s,xattr: %s,value: %s,flags: %d", name, s3Etag, s3MD5, 0)
+			return
```
Comment (davies) [inline pkg/gateway/gateway.go:892]: remove this

File: pkg/gateway/gateway.go [hunk 5]
```diff
@@ -666,7 +694,8 @@ func (n *jfsObjects) NewMultipartUpload(ctx context.Context, bucket string, obje
 }
 
 const uploadKeyName = "s3-object"
-const partEtag = "s3-etag"
+const s3PartEtag = "s3-part-etag"
```
Comment (davies) [inline pkg/gateway/gateway.go:697]: This is not compatible with older version, we should not change that.

File: pkg/gateway/gateway.go [hunk 6]
```diff
@@ -467,10 +468,22 @@ func (n *jfsObjects) CopyObject(ctx context.Context, srcBucket, srcObject, dstBu
...
 		return
 	}
+
+	var etag []byte
+	if n.keepEtag {
+		etag, _ = n.fs.GetXattr(mctx, src, s3Etag)
+		if string(etag) != "" {
+			eno = n.fs.SetXattr(mctx, dst, s3Etag, etag, 0)
+			if eno != 0 {
+				logger.Errorf("set xattr error, path: %s,xattr: %s,value: %s,flags: %d", dst, s3Etag, etag, 0)
```
Comment (davies) [inline pkg/gateway/gateway.go:478]: warning

File: pkg/gateway/gateway.go [hunk 7]
```diff
@@ -467,10 +468,22 @@ func (n *jfsObjects) CopyObject(ctx context.Context, srcBucket, srcObject, dstBu
 		err = jfsToObjectErr(ctx, eno, dstBucket, dstObject)
 		return
 	}
+
+	var etag []byte
+	if n.keepEtag {
+		etag, _ = n.fs.GetXattr(mctx, src, s3Etag)
+		if string(etag) != "" {
```
Comment (davies) [inline pkg/gateway/gateway.go:475]: len(etag) != 0

File: pkg/gateway/gateway.go [hunk 8]
```diff
@@ -530,13 +543,21 @@ func (n *jfsObjects) GetObjectInfo(ctx context.Context, bucket, object string, o
 		err = jfsToObjectErr(ctx, os.ErrNotExist, bucket, object)
 		return
 	}
+	var etag []byte
+	if n.keepEtag {
+		etag, eno = n.fs.GetXattr(mctx, n.path(bucket, object), s3Etag)
+		if eno != 0 {
+			logger.Errorf("get xattr error, path: %s,xattr: %s", n.path(bucket, object), s3Etag)
```
Comment (davies) [inline pkg/gateway/gateway.go:550]: It's OK to having no etag,  no logging here

File: pkg/gateway/gateway.go (hunk 5 above)
Comment (zhijian-pro) [reply pkg/gateway/gateway.go:697]: ok

**Review by davies — APPROVED** [review]

**Pull request merged**